samza-core/src/main/java/org/apache/samza/coordinator/metadatastore/CoordinatorStreamStore.java

```diff
@@ -0,0 +1,198 @@
+ *
+ * This class is thread safe.
+ */
+public class CoordinatorStreamStore implements MetadataStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CoordinatorStreamStore.class);
+  private static final String SOURCE = "SamzaContainer";
+
+  private final Config config;
+  private final SystemStream coordinatorSystemStream;
+  private final SystemStreamPartition coordinatorSystemStreamPartition;
+  private final SystemProducer systemProducer;
+  private final SystemConsumer systemConsumer;
+  private final SystemAdmin systemAdmin;
+  private final Object bootstrapLock = new Object();
+  private final String type;
+  private final Serde<List<?>> keySerde;
+
+  // Using custom comparator since java default comparator offers object identity equality(not value equality) for byte arrays.
+  private final Map<byte[], byte[]> bootstrappedMessages = new TreeMap<>(UnsignedBytes.lexicographicalComparator());
+  private final AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+  private SystemStreamPartitionIterator iterator;
+
+  public CoordinatorStreamStore(String namespace, Config config, MetricsRegistry metricsRegistry) {
+    this.config = config;
+    this.type = namespace;
+    this.keySerde = new JsonSerde<>();
+    this.coordinatorSystemStream = CoordinatorStreamUtil.getCoordinatorSystemStream(config);
+    this.coordinatorSystemStreamPartition = new SystemStreamPartition(coordinatorSystemStream, new Partition(0));
+    SystemFactory systemFactory = CoordinatorStreamUtil.getCoordinatorSystemFactory(config);
+    this.systemProducer = systemFactory.getProducer(this.coordinatorSystemStream.getSystem(), config, metricsRegistry);
+    this.systemConsumer = systemFactory.getConsumer(this.coordinatorSystemStream.getSystem(), config, metricsRegistry);
+    this.systemAdmin = systemFactory.getAdmin(this.coordinatorSystemStream.getSystem(), config);
+  }
+
+  @Override
+  public void init(Config config, MetricsRegistry metricsRegistry) {
+    if (isInitialized.compareAndSet(false, true)) {
+      LOG.info("Starting the coordinator stream system consumer with config: {}.", config);
+      registerConsumer();
+      systemConsumer.start();
+      systemProducer.register(SOURCE);
+      systemProducer.start();
+      iterator = new SystemStreamPartitionIterator(systemConsumer, coordinatorSystemStreamPartition);
+      bootstrapMessagesFromStream();
+    } else {
+      LOG.info("Store had already been initialized. Skipping.", coordinatorSystemStreamPartition);
+    }
+  }
+
+  @Override
+  public byte[] get(byte[] key) {
+    bootstrapMessagesFromStream();
+    return bootstrappedMessages.get(key);
+  }
+
+  @Override
+  public void put(byte[] key, byte[] value) {
+    try {
+      OutgoingMessageEnvelope envelope = new OutgoingMessageEnvelope(coordinatorSystemStream, 0, key, value);
+      systemProducer.send(SOURCE, envelope);
+      flush();
+    } catch (Exception e) {
+      throw new SamzaException(e);
```

Comment:
Don't need to wrap exceptions in SamzaException unless we have special handling for them at the call site. Here and everywhere else.

Reply:
Done.